Update fx proxy to use classmethod for __torch_function__

albanD opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #65145
* #64843
* __->__ #64842
* #64841

Change the `__torch_function__` to follow best guidelines of using classmethods.
I am not sure how to handle the case where multiple tracer objects are given as input but given that before we were getting an arbitrary tracer from there based on the "self" that was arbitrarily chosen by the torch_function caller, the new implementation is not worst?
Let me know what you think!

Differential Revision: [D30991423](https://our.internmc.facebook.com/intern/diff/D30991423)

albanD committed:
Update fx proxy to use classmethod for __torch_function__

[ghstack-poisoned]

albanD committed:
Update on "Update fx proxy to use classmethod for __torch_function__"


Change the `__torch_function__` to follow best guidelines of using classmethods.
I am not sure how to handle the case where multiple tracer objects are given as input but given that before we were getting an arbitrary tracer from there based on the "self" that was arbitrarily chosen by the torch_function caller, the new implementation is not worst?
Let me know what you think!

[ghstack-poisoned]

jamesr66a commented on torch/fx/proxy.py:
```diff
@@ -259,18 +260,29 @@ def __len__(self):
+            elif isinstance(a, collections.Sequence):
+                args_of_this_cls.extend(el for el in a if isinstance(el, cls))
+        assert len(args_of_this_cls) > 0
+        tracer = args_of_this_cls[0].tracer
```
I recommend instead using `map_aggregate` to recursively traverse `args` and `kwargs`:

https://github.com/pytorch/pytorch/blob/57eda692192ed4bcd291f09bf5b2e409c3031db6/torch/fx/node.py#L584

Something like:

```
tracers : Dict[torch.fx.Tracer, None] = []
def find_tracer(a):
  if isinstance(a, cls):
    proxy_inputs.setdefault(a.tracer)
torch.fx.node.map_aggregate(args, find_tracer)
torch.fx.node.map_aggregate(kwargs, find_tracer)

if len(tracers) > 1:
  raise RuntimeError(f'Found multiple different tracers {list(tracers.keys)} while trying to trace operations {orig_method}')
tracer = next(iter(tracers.keys()))
```

Multiple tracers would indeed be an error and we should just throw as in the example code

albanD replied:
Thanks for the pointer to map_aggregate, that makes it much cleaner.
Updated and fixed lint.

jamesr66a reviewed (approved):
LGTM! Recommend a change for how we traverse to find the tracer but otherwise I'm in support of this change

(also fix the lint issue)

albanD committed:
Update on "Update fx proxy to use classmethod for __torch_function__"


Change the `__torch_function__` to follow best guidelines of using classmethods.
I am not sure how to handle the case where multiple tracer objects are given as input but given that before we were getting an arbitrary tracer from there based on the "self" that was arbitrarily chosen by the torch_function caller, the new implementation is not worst?
Let me know what you think!

[ghstack-poisoned]

albanD committed:
Update on "Update fx proxy to use classmethod for __torch_function__"


Change the `__torch_function__` to follow best guidelines of using classmethods.
I am not sure how to handle the case where multiple tracer objects are given as input but given that before we were getting an arbitrary tracer from there based on the "self" that was arbitrarily chosen by the torch_function caller, the new implementation is not worst?
Let me know what you think!

[ghstack-poisoned]

albanD committed:
Update on "Update fx proxy to use classmethod for __torch_function__"


Change the `__torch_function__` to follow best guidelines of using classmethods.
I am not sure how to handle the case where multiple tracer objects are given as input but given that before we were getting an arbitrary tracer from there based on the "self" that was arbitrarily chosen by the torch_function caller, the new implementation is not worst?
Let me know what you think!

[ghstack-poisoned]

albanD committed:
Update on "Update fx proxy to use classmethod for __torch_function__"


Change the `__torch_function__` to follow best guidelines of using classmethods.
I am not sure how to handle the case where multiple tracer objects are given as input but given that before we were getting an arbitrary tracer from there based on the "self" that was arbitrarily chosen by the torch_function caller, the new implementation is not worst?
Let me know what you think!

[ghstack-poisoned]

albanD committed:
Update on "Update fx proxy to use classmethod for __torch_function__"


Change the `__torch_function__` to follow best guidelines of using classmethods.
I am not sure how to handle the case where multiple tracer objects are given as input but given that before we were getting an arbitrary tracer from there based on the "self" that was arbitrarily chosen by the torch_function caller, the new implementation is not worst?
Let me know what you think!

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30991423).

albanD committed:
Update on "Update fx proxy to use classmethod for __torch_function__"


Change the `__torch_function__` to follow best guidelines of using classmethods.
I am not sure how to handle the case where multiple tracer objects are given as input but given that before we were getting an arbitrary tracer from there based on the "self" that was arbitrarily chosen by the torch_function caller, the new implementation is not worst?
Let me know what you think!

Differential Revision: [D30991423](https://our.internmc.facebook.com/intern/diff/D30991423)

[ghstack-poisoned]

albanD commented:
@albanD has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30991423).

jamesr66a commented on torch/fx/proxy.py:
```diff
@@ -259,18 +259,32 @@ def __len__(self):
+
+        if len(tracers) > 1:
+            raise RuntimeError(f'Found multiple different tracers {list(tracers.keys())} while '
+                               'trying to trace operations {orig_method}')
```
I had to test this behavior because i didn't know what would happen but I think the continuation of the fstring needs the `f` sigil on it as well:

```
a = 3
b = 4

fstring = (f'i am a {a} '
            'fstring {b}')

print(fstring)
"""
i am a 3 fstring {b}
"""
```

albanD replied:
Ho that's interesting. Let update that.
